src/EntityComponentManager.cc

```diff
@@ -582,42 +611,77 @@ bool EntityComponentManager::CreateComponentImplementation(
   bool updateData = true;
 
   this->dataPtr->AddModifiedComponent(_entity);
-  this->dataPtr->entityComponents[_entity].insert(_componentTypeId);
   this->dataPtr->oneTimeChangedComponents[_componentTypeId].insert(_entity);
-  this->dataPtr->entityComponentsDirty = true;
 
   // Instantiate the new component.
   auto newComp = components::Factory::Instance()->New(_componentTypeId, _data);
 
-  auto compAddResult =
-    this->dataPtr->entityCompStorage.AddComponent(_entity, std::move(newComp));
-  switch (compAddResult)
+  // make sure the entity exists
+  auto typeMapIter = this->dataPtr->componentTypeIndex.find(_entity);
+  if (typeMapIter == this->dataPtr->componentTypeIndex.end())
+  {
+    ignerr << "Attempt to create a component of type [" << _componentTypeId
+      << "] attached to entity [" << _entity
+      << "] failed: entity not in componentTypeIndex." << std::endl;
+    return false;
+  }
+
+  auto entityCompIter = this->dataPtr->storage.find(_entity);
+  if (entityCompIter == this->dataPtr->storage.end())
   {
-    case ComponentAdditionResult::FAILED_ADDITION:
-      ignwarn << "Attempt to create a component of type [" << _componentTypeId
-        << "] attached to entity [" << _entity << "] failed.\n";
+    ignerr << "Attempt to create a component of type [" << _componentTypeId
+      << "] attached to entity [" << _entity
+      << "] failed: entity not in storage." << std::endl;
+    return false;
+  }
```

Comment:
minor optimization: should we place the `componentTypeIndex` and `storage` validity checks before the `components::Factory::Instance` call, so that we avoid creating the component if there's no entity for it in the storage?